pkg/kubelet/cm/dra/manager.go

```diff
@@ -108,48 +108,57 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			continue
 		}
 
-		claimInfo := m.cache.get(*claimName, pod.Namespace)
-		if claimInfo == nil {
-			// claim does not exist in cache, create new claimInfo object
-			// to be processed later.
-			claimInfo = newClaimInfoFromResourceClaim(resourceClaim)
-		}
+		// Atomically perform some operations on the claimInfo cache.
+		err = m.cache.withLock(func() error {
+			// Get a reference to the claim info for this claim from the cache.
+			// If there isn't one yet, then add it to the cache.
+			claimInfo, exists := m.cache.get(resourceClaim.Name, resourceClaim.Namespace)
+			if !exists {
+				claimInfo = m.cache.add(newClaimInfoFromClaim(resourceClaim))
+			}
 
-		// We delay checkpointing of this change until this call
-		// returns successfully. It is OK to do this because we
-		// will only return successfully from this call if the
-		// checkpoint has succeeded. That means if the kubelet is
-		// ever restarted before this checkpoint succeeds, the pod
-		// whose resources are being prepared would never have
-		// started, so it's OK (actually correct) to not include it
-		// in the cache.
-		claimInfo.addPodReference(pod.UID)
-
-		if claimInfo.prepared {
-			// Already prepared this claim, no need to prepare it again
-			continue
-		}
+			// Add a reference to the current pod in the claim info.
+			claimInfo.addPodReference(pod.UID)
 
-		// Loop through all plugins and prepare for calling NodePrepareResources.
-		for _, resourceHandle := range claimInfo.ResourceHandles {
-			// If no DriverName is provided in the resourceHandle, we
-			// use the DriverName from the status
-			pluginName := resourceHandle.DriverName
-			if pluginName == "" {
-				pluginName = resourceClaim.Status.DriverName
+			// Checkpoint to ensure the claim is tracked for this pod before preparing it.
+			if err := m.cache.syncToCheckpoint(); err != nil {
+				return fmt.Errorf("failed to checkpoint claimInfo state: %w", err)
 			}
```

Comment:
There's still a problem here...

The issue is that the pod reference will get checkpointed here even if the prepare call never succeeds. If the pod is then deleted before it *ever* has a successful prepare call, unprepare will never be called on it, and we will never remove a reference to it from the cache / checkpoint. Future pods that reference the claim will still be able to prepare it, but unprepare will be blocked since we have no mechanism to remove this now deleted pod from the cache / checkpoint.

One way to address this would be to:
1. Move the checkpoint to the end (as it was before)
2. Remove the pod reference from the cache if / when this function returns an error

Let me make the change on my end and push it.

Reply:
Pushed.